src/components/QueryEditor.tsx

```diff
@@ -1,56 +1,100 @@
   const [savedSearchIdOptions, setSavedSearchIdOptions] = useState<SelectableValue[]>([]);
+  const [savedSearchId, setSavedSearchId] = useState('');
 
-  const onSavedQueryIdChange = (sv: SelectableValue<string>) => {
-    const v = sv.value?.replace(/\s+/g, '') ?? ''; // auto-trim/remove any whitespace
-    if (v.match(/^[a-zA-Z0-9_]+$/)) {
-      onChange({ ...query, savedSearchId: v });
-      debouncedOnRunQuery();
+  const [queryType, setQueryType] = useState(query.type as string ?? DEFAULT_QUERY_TYPE);
+  const [adhocQuery, setAdhocQuery] = useState(query.type === 'adhoc' ? query.query : '');
+
+  const onQueryTypeChange = (sv: SelectableValue<string>) => {
+    const newQueryType = sv.value ?? DEFAULT_QUERY_TYPE;
+    setQueryType(newQueryType);
+    if (newQueryType === 'saved') {
+      onChange({ ...query, type: 'saved', savedSearchId });
+    } else {
+      onChange({ ...query, type: 'adhoc', query: adhocQuery });
+    }
+  };
+
+  const onAdhocQueryChange = (event: ChangeEvent<HTMLTextAreaElement>) => {
+    const newQuery = event.target.value;
+    setAdhocQuery(newQuery);
+    onChange({ ...query, type: 'adhoc', query: newQuery });
+    // Don't run it automatically, the user can hit Enter or click "Run query" when ready
+  };
+
+  const onAdhocQueryKeyDown = (event: KeyboardEvent) => {
+    if (event.key === 'Enter' && !event.shiftKey) { // allow shift-enter to add a line break
+      event.preventDefault();
+      if (adhocQuery.trim().length > 0) {
+        onRunQuery();
+      }
     }
   };
 
-  const onMaxResultsChange = (event: ChangeEvent<HTMLInputElement>) => {
-    onChange({ ...query, maxResults: Math.max(1, +event.target.value) });
-    debouncedOnRunQuery();
+  const onSavedQueryIdChange = (sv: SelectableValue<string>) => {
+    const newSavedSearchId = sv.value?.replace(/\s+/g, '') ?? ''; // auto-trim/remove any whitespace
+    setSavedSearchId(newSavedSearchId);
+    if (newSavedSearchId !== savedSearchId && newSavedSearchId.match(/^[a-zA-Z0-9_]+$/)) {
+      onChange({ ...query, type: 'saved', savedSearchId: newSavedSearchId });
+      debouncedOnRunQuery();
+    }
   };
 
   // Load the saved search IDs, let the user just pick one
   useEffect(() => {
     const loadSavedSearchIds = async () => {
       try {
         const savedSearchIds = await datasource.loadSavedSearchIds();
-        const options = [
-          { label: '', value: '' },
+        setSavedSearchIdOptions([
+          { label: 'Please select...', value: '' },
           ...savedSearchIds.map((value) => ({ value, label: value })),
-        ];
-        setSavedSearchIdOptions(options);
+        ]);
       } catch (err) {
         console.log(`Failed to load saved search IDs: ${err}`);
       }
     };
     loadSavedSearchIds();
   }, [datasource]);
 
+  const getQueryUI = () => {
+    if (queryType === 'saved') {
+      return <InlineField label="Saved Search" labelWidth={16} tooltip="ID of the Cribl saved search">
```

Comment:
Hate to make a formatting comment here, but can we please ensure this is wrapped in parens. Ideally, these kinds of issues are fixed by eslint/prettier. But, I don't think the repo is set up to format this nicely.)

```suggestion
      return (
        <InlineField label="Saved Search" labelWidth={16} tooltip="ID of the Cribl saved search">
```